test/util.ts

```diff
@@ -0,0 +1,119 @@
+import {Program} from "../src/runtime/dsl";
+import * as Runtime from "../src/runtime/runtime";
+import * as test from "tape";
+
+// You can specify changes as either [e,a,v] or [e,a,v,round,count];
+export type EAVArray = [Runtime.RawValue, Runtime.RawValue, Runtime.RawValue];
+export type EAVRCArray = [Runtime.RawValue, Runtime.RawValue, Runtime.RawValue, number, number];
+export type TestChange =  EAVArray | EAVRCArray;
+
+export function createChanges(transaction:number,eavns:TestChange[]) {
+  let changes:Runtime.Change[] = [];
+  for(let [e, a, v, round = 0, count = 1] of eavns as EAVRCArray[]) {
+    changes.push(Runtime.Change.fromValues(e, a, v, "my-awesome-node", transaction, round, count));
+  }
+  return changes;
+}
+
+export function verify(assert:any, program:Program, input:any[], output:any[], transaction = 1) {
+  let ins = createChanges(transaction, input);
+  let outs = createChanges(transaction, output);
+
+  let all = ins.concat(outs);
+  let {changes} = program.input(ins);
+  let msg = "Fewer changes than expected";
+  if(changes.length > all.length) {
+    msg = "More changes than expected";
+  }
+  assert.equal(changes.length, all.length, msg);
+
+  // Because the changes handed to us in expected aren't going to have the same
```

Comment:
I don't think this is greedy. It always keeps a list of all the people that could match, so if you have two possible matches that end up being exactly identical, meaning that every value of every field is the same at the same time (which I don't think is possible in our current semantics?), you could potentially end up not having resolved which is which, but at least proven that possibilities exist.

We only reduce the set of possibilities when we've discovered that some actual reduces the possible matches to only one. If we're worried about that happening to multiple somehow, we could keep a lookup of fully resolved guys and make sure the match isn't already taken. I *think* that'd prevent any false positives/negatives that could result from that weirdness.